infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesPersistentVolumeClaims.java

```diff
@@ -231,10 +218,56 @@ public void onClose(KubernetesClientException cause) {
       }
     } catch (KubernetesClientException e) {
       throw new KubernetesInfrastructureException(e);
-    } finally {
-      if (watch != null) {
-        watch.close();
-      }
     }
   }
+
+  private Watch pvcIsBoundWatcher(
+      CompletableFuture<PersistentVolumeClaim> future,
+      Resource<PersistentVolumeClaim, DoneablePersistentVolumeClaim> pvcResource,
+      Predicate<PersistentVolumeClaim> predicate,
+      String name) {
+    return pvcResource.watch(
+        new Watcher<PersistentVolumeClaim>() {
+          @Override
+          public void eventReceived(Action action, PersistentVolumeClaim pvc) {
+            if (predicate.test(pvc)) {
+              future.complete(pvc);
+            }
+          }
+
+          @Override
+          public void onClose(KubernetesClientException cause) {
+            future.completeExceptionally(
+                new InfrastructureException(
+                    "Waiting for persistent volume claim '" + name + "' was interrupted"));
+          }
+        });
+  }
+
+  private Watch pvcIsWaitingForConsumerWatcher(
+      CompletableFuture<PersistentVolumeClaim> future, PersistentVolumeClaim actualPvc)
+      throws InfrastructureException {
+    return clientFactory
+        .create(workspaceId)
+        .events()
+        .inNamespace(namespace)
+        .withField("reason", "WaitForFirstConsumer")
+        .withField("involvedObject.uid", actualPvc.getMetadata().getUid())
+        .watch(
+            new Watcher<Event>() {
+              private final Logger log = LoggerFactory.getLogger(getClass());
+
+              @Override
+              public void eventReceived(Action action, Event resource) {
+                log.debug(
+                    "PVC is waiting for first consumer. Don't wait to bound to avoid deadlock.");
+                future.complete(actualPvc);
+              }
+
+              @Override
+              public void onClose(KubernetesClientException cause) {
+                log.debug(cause.toString());
```

Comment:
Future is completed exceptionally in another listener on closing but not here. 
Should not they do in the same way?

Reply:
yes, fixed. I've added also check for null which means that it was closed normally (by try block)